### PR TITLE
Post Processing (Tonemapping + Color Correction)

### PR DESCRIPTION
**Pull request by Purrspctiv**
Implements tonemapping and color correction!
Also has some corrections to the sky shader.

KNOWN ISSUES:
- Corrupted LUT on specific maps. Known maps include cs_office and most hlvr maps. No clue why this is happening, this is the main thing to investigate and solve.
- Viewmodes are WIP

The code isn't fully cleaned up, and it's not ready to be merged into master yet. There's still some things I need to take care of. The main purpose of this pull request (currently) is so others can look into the corrupted LUT issue, which is completely baffling me.

### CODE REVIEW

File: ValveResourceFormat/Resource/ResourceTypes/PostProcessing.cs [hunk 1]
```diff
@@ -70,11 +70,11 @@ public byte[] GetRAWData()
             var j = 0;
             for (var i = 0; i < lut.Length; i++)
             {
-                // Skip each 4th byte
-                if (((i + 1) % 4) == 0)
-                {
-                    continue;
-                }
+                // Skip each 4th byte (this doesn't change anything)
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/PostProcessing.cs:73]: This makes the whole loop redundant. Does this still work correctly for the exporting?

File: GUI/Types/Renderer/ScenePostProcessVolume.cs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                    }
+                    Log.Info(nameof(ScenePostProcessVolume), pixelValues);
+                }
+#endif
+
+                var buffer = ArrayPool<byte>.Shared.Rent(dimensions * dimensions * dimensions * bytesPerPixel);
+
+                try
+                {
+                    GL.TextureSubImage3D(ColorCorrectionLUT.Handle, 0, 0, 0, 0, dimensions, dimensions, dimensions, PixelFormat.Rgba, PixelType.UnsignedByte, buffer);
```
Comment (xPaw) [inline GUI/Types/Renderer/ScenePostProcessVolume.cs:228]: A buffer is being rented, but nothing is written to it, so this writes garbage data to the texture (im surprised this worked at all on any map).

Did you mean to write `data` here without buffer?

Also keep in mind ArrayPool can return bigger sizes than requested.

File: GUI/Controls/GLViewerControl.cs [hunk 3]
```diff
@@ -570,19 +582,34 @@ private void HandleResize()
...
+            }
+
+            if (FinalFramebuffer.InitialStatus == FramebufferErrorCode.FramebufferUndefined)
+            {
+                var status = FinalFramebuffer.Initialize();
+
+                if (status != FramebufferErrorCode.FramebufferComplete)
+                {
+                    Log.Error(nameof(GLViewerControl), $"PostProcess Framebuffer failed to bind with error: {status}");
+                    Log.Info(nameof(GLViewerControl), "Falling back to default framebuffer.");
```
Comment (xPaw) [inline GUI/Controls/GLViewerControl.cs:609]: You can put this into the above message.

File: GUI/Types/Renderer/WorldLoader.cs [hunk 4]
```diff
@@ -712,6 +714,102 @@ static bool IsCamera(string cls)
...
+                        Transform = transformationMatrix, // needed if model is used
+                    };
+
+                    var postProcessResourceFilename = entity.GetProperty<string>("postprocessing");
+
+                    if (postProcessResourceFilename != null)
+                    {
+                        var postProcessResource = guiContext.LoadFileCompiled(postProcessResourceFilename);
+
+                        if (postProcessResource != null)
```
Comment (xPaw) [inline GUI/Types/Renderer/WorldLoader.cs:755]: Should we bail out and not call `AddPostProcessVolume` if LUT fails to load?

**Comment by xPaw** [comment]
Add a checkbox to disable color correction / post processing.

File: GUI/Types/Renderer/GLSceneViewer.cs [hunk 5]
```diff
@@ -184,6 +185,16 @@ public virtual void PreSceneLoad()
 
             var defaultCubeTexture = GuiContext.MaterialLoader.LoadTexture(cubeFogResource);
             Textures.Add(new(ReservedTextureSlots.FogCubeTexture, "g_tFogCubeTexture", defaultCubeTexture));
+
+
+            // TODO: add backup for this
```
Comment (xPaw) [inline GUI/Types/Renderer/GLSceneViewer.cs:190]: You can see how that's done for the BRDF texture.

File: GUI/Types/Renderer/Shaders/post_processing.frag [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+
+#if F_COLOR_CORRECTION_LUT
+
+layout(location=2) uniform sampler3D g_tColorCorrection;
+uniform float g_flColorCorrectionDefaultWeight;
+
+vec3 ApplyColorCorrection(vec3 vColor)
+{
+    vec3 ColorCorrectedColor = texture(g_tColorCorrection, saturate(vColor) * 0.9688 + 0.0156).rgb;
```
Comment (kristiker) [inline GUI/Types/Renderer/Shaders/post_processing.frag:84]:
```suggestion
    uniform int g_nColorCorrectionSize = 32;
    const float x = (g_nColorCorrectionSize-1/g_nColorCorrectionSize);
    const float y = (1/(g_nColorCorrectionSize*2);
    vec3 ColorCorrectedColor = texture(g_tColorCorrection, saturate(vColor) * x +y).rgb;
```

File: GUI/Types/Renderer/Shaders/post_processing.frag [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    // Finally, convert from Linear to Gamma space
+    vec3 vTonemappedColorSRGB = SrgbLinearToGamma(vTonemappedColor);
+
+    return vTonemappedColorSRGB;
+}
+
+
+
+#if F_COLOR_CORRECTION_LUT
```
Comment (kristiker) [inline GUI/Types/Renderer/Shaders/post_processing.frag:77]: I think we might be able to use this shader without any define, and use uniform bools if really necessary.

File: GUI/Controls/GLViewerControl.cs [hunk 8]
```diff
@@ -372,7 +374,8 @@ private static void OnDebugMessage(DebugSource source, DebugType type, int id, D
         private static readonly DebugProc OpenGLDebugMessageDelegate = OnDebugMessage;
 
         public Framebuffer GLDefaultFramebuffer;
-        public Framebuffer MainFramebuffer;
+        public Framebuffer BasePassFramebuffer;
+        public Framebuffer FinalFramebuffer;
```
Comment (kristiker) [inline GUI/Controls/GLViewerControl.cs:378]: Maybe we can postprocess directly from `MainFramebuffer` into the `GLDefaultFramebuffer`.

File: GUI/Types/Renderer/Shaders/post_processing.frag (hunk 6 above)
Comment (Purrspctiv) [reply GUI/Types/Renderer/Shaders/post_processing.frag:84]: Good point- funnily enough this is based on Valve's shader code, so I guess they just assumed it'll always be 32

File: GUI/Types/Renderer/Shaders/post_processing.frag (hunk 7 above)
Comment (Purrspctiv) [reply GUI/Types/Renderer/Shaders/post_processing.frag:77]: We'd have to provide a texture for g_tColorCorrection still though, right?

File: ValveResourceFormat/Resource/ResourceTypes/PostProcessing.cs (hunk 1 above)
Comment (Purrspctiv) [reply ValveResourceFormat/Resource/ResourceTypes/PostProcessing.cs:73]: Fixed!

File: GUI/Types/Renderer/Shaders/post_processing.frag (hunk 6 above)
Comment (xPaw) [reply GUI/Types/Renderer/Shaders/post_processing.frag:84]: That might have been optimized by the compiler.

**Comment by xPaw** [comment]
The vtex needs to be added as embedded resource here: https://github.com/ValveResourceFormat/ValveResourceFormat/blob/0486309021e6e3c875b42c94cc34d4598542754f/GUI/GUI.csproj#L41

File: GUI/Types/Renderer/Shaders/post_processing.frag (hunk 6 above)
Comment (kristiker) [reply GUI/Types/Renderer/Shaders/post_processing.frag]: Yeah weird that its hardcoded, but we can do better why not.

File: GUI/Types/Renderer/Shaders/post_processing.frag (hunk 7 above)
Comment (kristiker) [reply GUI/Types/Renderer/Shaders/post_processing.frag:60]: With a boolean check, maybe not.

(now I'm wondering what a passthrough LUT would be like lol)

File: GUI/Types/Renderer/Shaders/post_processing.frag [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    // So, we apply a 1-color-value dither to break up color banding.
+    // This is part of the original code and actually works extremely well.
+    vec3 subPrecisionDither = (blueNoise - 0.5) * g_vBlueNoiseDitherParams.w;
+    return vColor + subPrecisionDither;
+}
+
+vec4 SampleColorBuffer(vec2 coords)
+{
+    vec4 singleSampleColor = texelFetch(g_tColorBuffer, ivec2(coords.xy), int(gl_SampleID));
+    singleSampleColor.rgb = singleSampleColor.rgb / (max3(singleSampleColor.rgb) + 1.0);
```
Comment (kristiker) [inline GUI/Types/Renderer/Shaders/post_processing.frag:74]: What's the divide by `max3(singleSampleColor.rgb) +1 ` doing

File: GUI/Types/Renderer/ScenePostProcessVolume.cs (hunk 2 above)
Comment (kristiker) [reply GUI/Types/Renderer/ScenePostProcessVolume.cs]: > im surprised this worked at all on any map

I'm guessing it rented same array as the one rented by binarykv3.

**Pull request merged**